### PR TITLE
[MRG]:sparkles: Update register information (metric and task) and add an object detection example

### LINKED ISSUE #166: [Feat] support single-stage detection model
- [ ] https://github.com/Media-Smart/vedadet
- [x] check whether the converter can support this
- [x] improve the converter
- [ ] improve the profiler
- [ ] improve the deployer
- [x] add a new notebook

### PR DESCRIPTION
**Pull request by univerone**
Closes #166 #167 

#### What has been done to verify that this works as intended?

- add Metric and Task enum class, modify the parameters of `register_model` function
- add a `model_input` parameter enables users to use a self-defined model input tensor
- modify related usages in documents

#### Does this change require updates to documentation?
Yes
#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

File: modelci/types/bo/model_objects.py [hunk 1]
```diff
@@ -47,6 +56,15 @@ class Status(Enum):
     FAIL = 3
 
 
+@unique
+class Metric(Enum):
+    """numerator of model evaluation metric
+    """
+    ACC = 0
```
Comment (HuaizhengZhang) [inline modelci/types/bo/model_objects.py:63]: better add comments
acc：for image classification
Map：for xxx

File: modelci/hub/converter.py [hunk 2]
```diff
@@ -229,27 +233,31 @@ def from_torch_module(
...
             input_names=input_names,  # the model's input names
-            output_names=['output'],  # the model's output names
+            output_names=output_names,  # the model's output names
             keep_initializers_as_inputs=True,
             **export_kwargs
         )
 
         if optimize:
-            network = ONNXConverter.optim_onnx(save_path_with_ext)
+            onnx_model = onnx.load(str(save_path_with_ext))
```
Comment (HuaizhengZhang) [inline modelci/hub/converter.py:259]: has this been tested？
which onnx version do you use？

File: example/resnet50_explicit_path.yml [hunk 3]
```diff
@@ -10,6 +10,7 @@ outputs:
     dtype: TYPE_FP32
 architecture: ResNet50
 framework: PYTORCH
+task: IMAGE_CLASSIFICATION
```
Comment (HuaizhengZhang) [inline example/resnet50_explicit_path.yml:13]: task is a static label. Move it to the third line better

**Comment by HuaizhengZhang** [comment]
@YuanmingLeee Please double check.

File: example/notebook/object_detection_model_deployment.ipynb [hunk 4]
```diff
@@ -0,0 +1,480 @@
+{
```
Comment (HuaizhengZhang) [inline example/notebook/object_detection_model_deployment.ipynb:1]: image
I cannot open the file. Is anything wrong?

File: modelci/hub/converter.py (hunk 2 above)
Comment (univerone) [reply modelci/hub/converter.py:259]: Yes, it has been tested, this is because `ONNXConverter.optim_onnx` only receives `onnx.ModelProto` object as input instead of PosixPath, It's a fix.

https://github.com/cap-ntu/ML-Model-CI/blob/8165e367ae14659ef27843e8774084b640ef0a76/modelci/hub/converter.py#L307

File: example/notebook/object_detection_model_deployment.ipynb (hunk 4 above)
Comment (univerone) [reply example/notebook/object_detection_model_deployment.ipynb:1]: https://github.com/univerone/ML-Model-CI/blob/detection/example/notebook/object_detection_model_deployment.ipynb

I can open it normally, you can try again on this link

File: example/resnet50.yml [hunk 5]
```diff
@@ -1,7 +1,8 @@
-weight: /home/lym/.modelci/ResNet50/pytorch-torchscript/1.zip
+weight: ~/.modelci/ResNet50/pytorch-torchscript/image_classification/1.zip
```
Comment (YuanmingLeee) [inline example/resnet50.yml:1]: Have you tried to register with this yaml file? I don't think the user-relative path `~/.modelci/...` will be parsed properly by Python

File: example/resnet50_explicit_path.yml [hunk 6]
```diff
@@ -1,7 +1,8 @@
-weight: /home/lym/.modelci/ResNet50/pytorch-torchscript/1.zip
+weight: ~/.modelci/ResNet50/pytorch-torchscript/image_classification/1.zip
```
Comment (YuanmingLeee) [inline example/resnet50_explicit_path.yml:1]: Same problem as stated above

File: modelci/hub/converter.py [hunk 7]
```diff
@@ -189,6 +189,8 @@ def from_torch_module(
             model: torch.nn.Module,
             save_path: Path,
             inputs: Iterable[IOShape],
+            outputs: Iterable[IOShape],
+            model_input: Optional[List] = None,
```
Comment (YuanmingLeee) [inline modelci/hub/converter.py:193]: Add a todo, we will reuse `inputs` to pass in `model_input` value later.

File: modelci/hub/manager.py [hunk 8]
```diff
@@ -262,7 +290,7 @@ def get_remote_model_weight(model: ModelBO):
         1. set force insert config.pbtxt
         2. set other options in generation of config.pbtxt (e.g. max batch size, instance group...)
     This function will keep a local cache of the used model in the path:
-        `~/.modelci/<architecture_name>/<framework>-<engine>/version`
+        `~/.modelci/<architecture_name>/<framework>-<engine>/<task>-<version>`
```
Comment (YuanmingLeee) [inline modelci/hub/manager.py:293]: this documentation is inconsistent with the path you are saving to.

**Review by YuanmingLeee — COMMENTED** [review]
This is a breaking MR, you should also test if the new saving dir can be used for model deployment.
I feel that the auto-deployment module (https://github.com/cap-ntu/ML-Model-CI/blob/master/modelci/hub/deployer/dispatcher.py) will fail.

File: example/resnet50.yml (hunk 5 above)
Comment (univerone) [reply example/resnet50.yml:1]: I change home directory to `~` to avoid personal information, we could use `/home/modelci/` instead

Comment (univerone) [reply example/resnet50.yml:1]: We could use `PosixPath` instead of `Path` when parsing this location to resolve this problem

Comment (YuanmingLeee) [reply example/resnet50.yml:1]: 1. It is good to use `<your home dir>`.

2. `PosixPath` does not understand `~`. You can try this:
```python
from pathlib import PosixPath
p = PosixPath('~')
print(p.absolute())
# PosixPath('/<home dir>/Documents/PyCharmProjects/ml-cloud-autoscaler/~')
```

Comment (univerone) [reply example/resnet50.yml:1]: How about use `os.path.expanduser` to wrap the path : https://docs.python.org/3/library/os.path.html#os.path.expanduser

Comment (YuanmingLeee) [reply example/resnet50.yml:1]: Nice! You can mark it as a TODO, and open another PR to solve it. Now it is ok to leave it as `<your home dir>` or `/home/modelci` for demostration.

Comment (univerone) [reply example/resnet50.yml:1]: OK!

File: example/resnet50_explicit_path.yml (hunk 6 above)
Comment (univerone) [reply example/resnet50_explicit_path.yml:1]: > This is a breaking MR, you should also test if the new saving dir can be used for model deployment.
> I feel that the auto-deployment module (https://github.com/cap-ntu/ML-Model-CI/blob/master/modelci/hub/deployer/dispatcher.py) will fail.

I have run the following code 

```python
    from torchvision import models
    model = models.resnet50(pretrained=True)
    from modelci.hub.manager import register_model
    from modelci.types.bo import Framework, IOShape, ModelVersion, Engine, Metric, Task
    from modelci.types.trtis_objects import ModelInputFormat
    # set model input and output formats
    inputs = [IOShape([-1, 3, 224, 224], dtype=float, name='INPUT__0', format=ModelInputFormat.FORMAT_NCHW)]
    outputs = [IOShape([-1, 1000], dtype=float, name='probs')]
    # register the model
    register_model(
        model,
        dataset='imagenet',
        metric={Metric.ACC: 0.76},
        task=Task.IMAGE_CLASSIFICATION,
        inputs=inputs,
        outputs=outputs,
        architecture='ResNet50',
        framework=Framework.PYTORCH,
        version=ModelVersion('2')
    )
    model_info = retrieve_model(architecture_name='ResNet50', framework=Framework.PYTORCH, engine=Engine.TORCHSCRIPT)
    # deploy the model to cuda device 0.
    serve(save_path=model_info[0].saved_path, device='cuda:0', name='torchscript-serving', batch_size=16)
```

and can return a running container, I have tested both `parse_path` and  `generate_path` function before.

Comment (YuanmingLeee) [reply example/resnet50_explicit_path.yml]: LGTM

**Review by YuanmingLeee — APPROVED** [review]
LGTM
